src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs

```diff
@@ -19,46 +21,82 @@ public DefaultTagHelperResolver(bool designTime)
         public override IReadOnlyList<TagHelperDescriptor> GetTagHelpers(Compilation compilation)
         {
             var results = new List<TagHelperDescriptor>();
+            var errors = new ErrorSink();
+
+            VisitTagHelpers(compilation, results, errors);
+            VisitViewComponents(compilation, results, errors);
+
+            return results;
+        }
 
-            // If ITagHelper isn't defined, then we couldn't possibly find anything.
+        private void VisitTagHelpers(Compilation compilation, List<TagHelperDescriptor> results, ErrorSink errors)
+        {
             var @interface = compilation.GetTypeByMetadataName(TagHelperTypes.ITagHelper);
             if (@interface == null)
             {
-                return results;
+                // If ITagHelper isn't defined, then we couldn't possibly find anything.
+                return;
             }
 
             var types = new List<INamedTypeSymbol>();
-            var visitor = new Visitor(@interface, types);
+            var visitor = new TagHelperVisitor(@interface, types);
 
-            visitor.Visit(compilation.Assembly.GlobalNamespace);
-
-            foreach (var reference in compilation.References)
-            {
-                if (compilation.GetAssemblyOrModuleSymbol(reference) is IAssemblySymbol assembly)
-                {
-                    visitor.Visit(assembly.GlobalNamespace);
-                }
-            }
+            VisitCompilation(visitor, compilation);
 
-            var errors = new ErrorSink();
             var factory = new DefaultTagHelperDescriptorFactory(compilation, DesignTime);
 
             foreach (var type in types)
             {
                 var descriptors = factory.CreateDescriptors(type, errors);
                 results.AddRange(descriptors);
             }
+        }
 
-            return results;
+        private static void VisitViewComponents(Compilation compilation, List<TagHelperDescriptor> results, ErrorSink errors)
+        {
+            var @interface = compilation.GetTypeByMetadataName(ViewComponentTypes.ViewComponentAttribute);
+            var types = new List<INamedTypeSymbol>();
+            var visitor = new ViewComponentVisitor(@interface, types);
+
+            VisitCompilation(visitor, compilation);
+
+            var factory = new ViewComponentTagHelperDescriptorFactory(compilation);
+
+            foreach (var type in types)
+            {
+                try
+                {
+                    var descriptor = factory.CreateDescriptor(type);
```

Comment:
These bits will be replaced by MVC invocations. Didn't feel that it was worth the time to convert each exception clause into an `errorSink.OnError` call/flow the ErrorSink throughout the factory.